src/ChunkedDecoder.php

```diff
@@ -0,0 +1,157 @@
+
+        $this->emit('close');
+        $this->removeAllListeners();
+    }
+
+    /** @internal */
+    public function handleEnd()
+    {
+        if (!$this->closed) {
+            $this->handleError(new \Exception('Unexpected `end` event'));
+        }
+    }
+
+    /** @internal */
+    public function handleError(\Exception $e)
+    {
+        $this->emit('error', array($e));
+        $this->close();
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        while ($this->buffer !== '') {
+            if (!$this->headerCompleted) {
+                $positionClrf = strpos($this->buffer, static::CRLF);
+
+                if ($positionClrf === false) {
+                    // Header shouldn't be bigger than 1024 bytes
+                    if (isset($this->buffer[1024])) {
+                        $this->handleError(new \Exception('Chunk size inclusive extension bigger than 1024 bytes'));
+                    }
+                    return;
+                }
+
+                $header = strtolower((string)substr($this->buffer, 0, $positionClrf));
+                $hexValue = $header;
+
+                if (strpos($header, ';') !== false) {
+                    $array = explode(';', $header);
+                    $hexValue = $array[0];
+                }
+
+                $this->chunkSize = hexdec($hexValue);
+                if (dechex($this->chunkSize) !== $hexValue || $hexValue < 0) {
+                    $this->handleError(new \Exception($hexValue . ' is not a valid hexadecimal number'));
+                    return;
+                }
+
+                $this->buffer = (string)substr($this->buffer, $positionClrf + 2);
+                $this->headerCompleted = true;
+                if ($this->buffer === '') {
+                    return;
+                }
+            }
+
+            $chunk = (string)substr($this->buffer, 0, $this->chunkSize - $this->transferredSize);
+
+            if ($chunk !== '') {
+                $this->transferredSize += strlen($chunk);
+                $this->emit('data', array($chunk));
+                $this->buffer = (string)substr($this->buffer, strlen($chunk));
+            }
+
+            if (strpos($this->buffer, static::CRLF) !== false) {
+                if ($this->chunkSize === 0) {
+                    $this->emit('end');
+                    $this->close();
+                    return;
+                }
+                $this->chunkSize = 0;
+                $this->headerCompleted = false;
+                $this->transferredSize = 0;
+                $this->buffer = (string)substr($this->buffer, 2);
+            }
+
+            if ($this->headerCompleted) {
+                if (strlen($this->buffer) > 2 && $this->chunkSize === $this->transferredSize) {
+                    // Send error event, the first 2 characters should be CLRF
+                    $this->handleError(new \Exception('Chunk does not end with a CLRF'));
```

Comment:
I can't comment the previous comment, but: Yes `4\r\ntestWOOT\r\n` is handled. See tests `testNoCrlfInChunk` and `testNoCrlfInChunkSplitted` in `ChunkedDecoderTests`.

To this comment. This has to be seen in relation to the previous if-statement. If the previous couldn't statement couldn't find a CLRF, this will still buffered. But only if the isn't string length of the current buffer isn't >2. That would mean that the data in the buffer isn't correct chunked encoding.